pkg/controller/registryinfra/registryinfra_controller.go

```diff
@@ -0,0 +1,795 @@
+		return
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+
+	<-stopCh
+}
+
+func (c *Controller) enqueue(cluster metav1.Object) {
+	key, err := controller.KeyFunc(cluster)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %#v: %v", cluster, err))
+		return
+	}
+
+	c.queue.Add(key)
+}
+
+// worker runs a worker thread that just dequeues items, processes them, and marks them done.
+// It enforces that the syncHandler is never invoked concurrently with the same key.
+func (c *Controller) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *Controller) processNextWorkItem() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+
+	err := c.syncHandler(key.(string))
+	c.handleErr(err, key)
+
+	return true
+}
+
+func (c *Controller) handleErr(err error, key interface{}) {
+	if err == nil {
+		c.queue.Forget(key)
+		return
+	}
+
+	logger := c.logger.WithField("cluster", key)
+
+	logger.Errorf("error syncing registry infra: %v", err)
+	if c.queue.NumRequeues(key) < maxRetries {
+		logger.Errorf("retrying registry infra")
+		c.queue.AddRateLimited(key)
+		return
+	}
+
+	utilruntime.HandleError(err)
+	logger.Infof("dropping cluster out of the queue: %v", err)
+	c.queue.Forget(key)
+}
+
+func (c *Controller) syncCluster(key string) error {
+	startTime := time.Now()
+	c.logger.WithField("key", key).Debug("syncing registry infra")
+	defer func() {
+		c.logger.WithFields(log.Fields{
+			"key":      key,
+			"duration": time.Now().Sub(startTime),
+		}).Debug("finished syncing registry infra")
+	}()
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	capiCluster, err := c.clusterLister.Clusters(namespace).Get(name)
+	if errors.IsNotFound(err) {
+		c.logger.WithField("key", key).Debug("cluster has been deleted")
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	cluster, err := controller.ConvertToCombinedCluster(capiCluster)
+	if err != nil {
+		return fmt.Errorf("failure converting to combined cluster: %v", err)
+	}
+
+	//clusterProviderStatus, err := controller.ClusterProviderStatusFromCluster(capiCluster)
+	_, err = controller.ClusterProviderStatusFromCluster(capiCluster)
+	if err != nil {
+		return fmt.Errorf("error retrieving clusterproviderstatus: %v", err)
+	}
+
+	// check if the cluster is being deleted
+	if capiCluster.DeletionTimestamp != nil {
+		//if true {
+		c.logger.Debugf("cleaning up registry infrastructure pieces for cluster %v", capiCluster.Name)
+
+		err = c.syncRegistryInfra(cluster, false /* deprovision infra */)
+		if err != nil {
+			return fmt.Errorf("error deleting registry infra: %v", err)
+		}
+		err = c.deleteFinalizer(capiCluster)
+		return err
+	}
+
+	if !hasRegistryInfraFinalizer(capiCluster) {
+		c.logger.Debugf("adding registryinfra finalizer to cluster %v", capiCluster.Name)
+		return c.addFinalizer(capiCluster)
+	}
+
+	if true {
```

Comment:
So your options would be to use the generation of the cluster you started tracking above, so we re-do this any time the cluster changed, or do what the awselb controller does and re-do it after some period of time has transpired (by tracking the last time we setup the registry infra in the status). I'd vote for the latter as these both feel very similar, and we'd like them to periodically make sure this is in place and healthy, just not on every single sync.